File: app/src/main/java/rod/bailey/trafficatnsw/hazard/details/HazardDetailsListAdapter.java
package rod.bailey.trafficatnsw.hazard.details;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListAdapter;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

import rod.bailey.trafficatnsw.hazard.HazardListItemView;
import rod.bailey.trafficatnsw.json.hazard.XArrangementElement;
import rod.bailey.trafficatnsw.json.hazard.XHazard;
import rod.bailey.trafficatnsw.json.hazard.XLane;
import rod.bailey.trafficatnsw.json.hazard.XPeriod;
import rod.bailey.trafficatnsw.json.hazard.XRoad;
import rod.bailey.trafficatnsw.util.DateUtils;
import rod.bailey.trafficatnsw.util.ListHeadingView;

public class HazardDetailsListAdapter extends BaseAdapter implements
		ListAdapter {

	private List<CellRec> cellRecs = new LinkedList<CellRec>();

	private final XHazard hazard;

	public HazardDetailsListAdapter(XHazard hazard) {
		assert hazard != null;
		this.hazard = hazard;

		// Title cell (always appears)
		addTitleCellRec();

		// "WHEN" heading
		addHeadingCellRec("WHEN");

		// "Cretaed" and "Last update" cells
		addCreateAndLastUpdatedCellRecs();

		// "Duration"
		addDurationCellRec();

		// "GENERAL" heading
		addHeadingCellRec("GENERAL");

		// "Attending"
		addAttendingCellRec();

		// "RTA Advice" cell
		addRtaAdviceCellRec();

		// "Queues" celll
		addQueueLengthCellRec();

		// "Closures" cells
		addClosureCellRecs();

		// "XLane" cells (each is an instance of LineCellRec in the lineCellRec
		// aray
		addLaneCellRecs();

		// "Traffic volume" and "Delay"
		addTrafficVolumeAndDelayCellRec();

		// "Public transport" hs HTML
		addPublicTransportCellRec();

		// "Arrangement elements" have HTML each one
		addArrangementElementCellRecs();

		// "Other advice" has HTML
		addOtherAdviceCellRec();

		removeFirstOfConsecutiveHeadings();
	}

	@Override
	public boolean isEnabled(int position) {
		return false;
	}
	
	private void addArrangementElementCellRecs() {
		if (!isEmptyList(hazard.getArrangementElements())) {
			for (XArrangementElement element : hazard.getArrangementElements()) {
				addHeadingCellRec(element.getTitle());
				addHtmlFieldCellRec(element.getHtml());
			}
		}
	}

	private void addAttendingCellRec() {
		// "Attending" row
		if (!isEmptyList(hazard.getAttendingGroups())) {
			int numAttendingGroups = hazard.getAttendingGroups().size();

			StringBuffer str = new StringBuffer();

			for (int i = 0; i < numAttendingGroups; i++) {
				String group = hazard.getAttendingGroups().get(i);

				if (!isEmptyStr(group)) {
					str.append(group);
				}

				if (i < (numAttendingGroups - 1)) {
					str.append(", ");
				}
			} // for i

			String strstr = str.toString();
			if (!isEmptyStr(strstr)) {
				TextFieldCellRec cellrec = new TextFieldCellRec("Attending",
						strstr);
				cellRecs.add(cellrec);
			}
		}
	}

	private void addClosureCellRecs() {
		if (hazard.getPeriods() != null) {
			for (XPeriod period : hazard.getPeriods()) {
				if ("ROAD_CLOSURE".equals(period.getClosureType())
						|| "LANE_CLOSURE".equals(period.getClosureType())) {
					// Planned road closure of lane closure

					StringBuffer str = new StringBuffer(period.getFromDay());

					if (!isEmptyStr(period.getToDay())) {
						str.append(" to ");
						str.append(period.getToDay());
					}

					if (!isEmptyStr(period.getStartTime())) {
						str.append(" (");
						str.append(period.getStartTime());

						if (!isEmptyStr(period.getFinishTime())) {
							str.append(" to ");
							str.append(period.getFinishTime());
						}

						str.append(")");
					}

					if (!isEmptyStr(period.getDirection())) {
						str.append(". ");
						str.append(period.getDirection());
						str.append(" affected");
					}

					String fieldValue = str.toString();
					String fieldName = null;

					if ("ROAD_CLOSURE".equals(period.getClosureType())) {
						fieldName = "Roads closed";
					} else {
						fieldName = "Lanes closed";
					}

					TextFieldCellRec cellRec = new TextFieldCellRec(fieldName,
							fieldValue);
					cellRecs.add(cellRec);
				}
			}
		}
	}

	private void addCreateAndLastUpdatedCellRecs() {
		// First CellRec is for "Created" date/time
		if (hazard.getCreated() != null) {
			String createdStr = DateUtils.relativeDateAndTime(
					hazard.getCreated(), true);
			TextFieldCellRec createdCellRec = new TextFieldCellRec("Started",
					createdStr);
			cellRecs.add(createdCellRec);
		}

		// Second CellRec is for "Last updated" date/time. But we only create a
		// 'last updated' cellrec if it will show a different date to the
		// 'created' date.
		if (!isLastUpdatedSameAsCreated()) {
			String lastUpdatedStr = DateUtils.relativeDateAndTime(
					hazard.getLastUpdated(), true);
			TextFieldCellRec lastUpdatedCellRec = new TextFieldCellRec(
					"Last checked", lastUpdatedStr);
			cellRecs.add(lastUpdatedCellRec);
		}
	}

	private void addDurationCellRec() {
		if ((hazard.getStart() != null) && (hazard.getEnd() != null)) {
			StringBuffer formatString = new StringBuffer("EEE d MMM");
			SimpleDateFormat sdf = new SimpleDateFormat(formatString.toString());

			String startStr = sdf.format(hazard.getStart());
			String endStr = sdf.format(hazard.getEnd());

			StringBuffer startToEnd = new StringBuffer();
			startToEnd.append(startStr);
			startToEnd.append(" - ");
			startToEnd.append(endStr);

			TextFieldCellRec cellrec = new TextFieldCellRec("Duration",
					startToEnd.toString());
			cellRecs.add(cellrec);
		}
	}

	private void addHeadingCellRec(String heading) {
		cellRecs.add(new HeadingCellRec(heading));
	}

	private void addHtmlFieldCellRec(String html) {
		HtmlFieldCellRec rec = new HtmlFieldCellRec(html, null, 0);
		cellRecs.add(rec);
	}

	private void addLaneCellRecs() {
		if (!isEmptyList(hazard.getRoads())) {
			XRoad firstRoad = hazard.getRoads().get(0);

			if (!isEmptyList(firstRoad.getImpactedLanes())) {
				for (XLane lane : firstRoad.getImpactedLanes()) {
					if (!isEmptyStr(lane.getExtent())) {
						String line = null;

						if ("Affected".equals(lane.getExtent())) {
							if ("Both directions".equals(lane
									.getAffectedDirection())) {
								line = "Traffic in both directions is affected.";
							} else {
								line = String.format("%s traffic is affected.",
										lane.getAffectedDirection());
							}
						} else if ("Lanes closed".equals(lane.getExtent())) {
							StringBuffer str = new StringBuffer();

							// Form a string of form
							// "X of Y <DIR>bound lanes closed"
							// e.g. "2 of 3 eastbound lanes closed".
							// NOTE: Special case where X and Y are blank
							// implies that both
							// X and Y are "1", but instead of saying
							// "1 of 1 <DIR>bound lanes closed"
							// we say "<DIR>bound lane closed".
							if (isEmptyStr(lane.getClosedLanes())
									&& isEmptyStr(lane.getNumberOfLanes())) {
								str.append(lane.getAffectedDirection());
								str.append(" lane closed");
							} else {
								str.append(lane.getClosedLanes());
								str.append(" of ");
								str.append(lane.getNumberOfLanes());
								str.append(" ");
								str.append(lane.getAffectedDirection()
										.toLowerCase());
								str.append(" lanes closed");
							}

							// Suffix "(<description>)" e.g. "(right turn lane)"
							if (!isEmptyStr(lane.getDescription())) {
								str.append(" (");
								str.append(lane.getDescription().toLowerCase());
								str.append(")");
							}

							str.append(".");
							LineCellRec rec = new LineCellRec(str.toString());
							cellRecs.add(rec);
						} else if ("Closed".equals(lane.getExtent())) {
							StringBuffer str = new StringBuffer();
							str.append(lane.getRoadType());
							str.append(" closed in ");

							if ("Both directions".equals(lane
									.getAffectedDirection())) {
								str.append("both directions.");
							} else {
								str.append(lane.getAffectedDirection()
										.toLowerCase());
								str.append(" direction.");
							}

							LineCellRec rec = new LineCellRec(str.toString());
							cellRecs.add(rec);
						}
					}
				}
			}
		}
	}

	private void addOtherAdviceCellRec() {
		if (!isEmptyStr(hazard.getOtherAdvice())) {
			addHeadingCellRec("Other advice");
			addHtmlFieldCellRec(hazard.getOtherAdvice());
		}
	}

	private void addPublicTransportCellRec() {
		if (!isEmptyStr(hazard.getPublicTransport())) {
			addHeadingCellRec("Public transport");
			addHtmlFieldCellRec(hazard.getPublicTransport());
		}
	}

	private void addQueueLengthCellRec() {
		if (!isEmptyList(hazard.getRoads())) {
			XRoad firstRoad = hazard.getRoads().get(0);

			if ((firstRoad != null) && (firstRoad.getQueueLength() > 0)) {
				String fieldName = "Queues";
				String fieldValue = null;

				StringBuffer queueStr = new StringBuffer();
				int queueLengthNumber = firstRoad.getQueueLength();

				queueStr.append("" + queueLengthNumber);
				queueStr.append("km");

				TextFieldCellRec cellrec = new TextFieldCellRec(fieldName,
						fieldValue);
				cellRecs.add(cellrec);
			}
		}
	}

	private void addRtaAdviceCellRec() {
		if (!isEmptyStr(hazard.getAdviceA())
				|| !isEmptyStr(hazard.getAdviceB())) {
			StringBuffer value = new StringBuffer();

			String fieldName = "Advice";

			if (!isEmptyStr(hazard.getAdviceA())) {
				value.append(hazard.getAdviceA());

				if (!isEmptyStr(hazard.getAdviceB())) {
					value.append(", ");
				}
			}

			if (!isEmptyStr(hazard.getAdviceB())) {
				value.append(hazard.getAdviceB());
			}

			TextFieldCellRec cellRec = new TextFieldCellRec(fieldName,
					value.toString());
			cellRecs.add(cellRec);
		}
	}

	private void addTitleCellRec() {
		cellRecs.add(new TitleCellRec(hazard));
	}

	/**
	 * "Traffic Volume" and "Delay" - adds a single LineCellRec containing a
	 * summary of both like "Heavy traffic. Significan delays" where the stings
	 * "Heavy" and "Significant" are taken from the JSON file.
	 */
	private void addTrafficVolumeAndDelayCellRec() {
		if (!isEmptyList(hazard.getRoads())) {
			XRoad firstRoad = hazard.getRoads().get(0);

			StringBuffer str = new StringBuffer();

			if (!isEmptyStr(firstRoad.getTrafficVolume())) {
				str.append(firstRoad.getTrafficVolume());
				str.append(" traffic.");
			}

			if (!isEmptyStr(firstRoad.getDelay())) {
				if (str.length() > 1) {
					str.append(" ");
				}

				str.append(firstRoad.getDelay());
				str.append(" delays.");
			}

			if (str.length() > 1) {
				LineCellRec rec = new LineCellRec(str.toString());
				cellRecs.add(rec);
			}
		}
	}

	private View cellForHtmlCellRecAtIndex(int i) {
		return null;
	}

	private View cellForNonHtmlCellRecAtIndex(int cellRecIndex) {
		return null;
	}

	private HtmlFieldCellRec createHtmlFieldCellRec(String fieldName,
			String fieldValue, int tag) {
		return null;
	}

	@Override
	public int getCount() {
		return cellRecs.size();
	}

	private double getHeightForHtmlCellRecAtIndex(int index) {
		return 0;
	}

	@Override
	public Object getItem(int position) {
		return cellRecs.get(position);
	}

	@Override
	public long getItemId(int position) {
		return position;
	}

	@Override
	public View getView(int position, View convertView, ViewGroup parent) {
		View result = null;
		Context ctx = parent.getContext();

		CellRec cellRec = cellRecs.get(position);

		if (cellRec instanceof TitleCellRec) {
			TitleCellRec titleCellRec = (TitleCellRec) cellRec;
			result = new HazardListItemView(ctx, titleCellRec.hazard, false,
					false);
		} else if (cellRec instanceof HeadingCellRec) {
			HeadingCellRec headingCellRec = (HeadingCellRec) cellRec;
			result = new ListHeadingView(ctx, headingCellRec.heading, false);
		} else if (cellRec instanceof LineCellRec) {
			LineCellRec lineCellRec = (LineCellRec) cellRec;
			result = new LineListItemView(ctx, lineCellRec.getLine());
		} else if (cellRec instanceof TextFieldCellRec) {
			TextFieldCellRec textCellRec = (TextFieldCellRec) cellRec;
			result = new TextFieldListItemView(ctx, textCellRec.fieldName,
					textCellRec.fieldValue);
		} else if (cellRec instanceof HtmlFieldCellRec) {
			HtmlFieldCellRec htmlCellRec = (HtmlFieldCellRec) cellRec;
			result = new HtmlListItemView(ctx, htmlCellRec.fieldHtml);
		}

		return result;
	}

	private boolean isEmptyList(List list) {
		boolean result = false;

		if (list == null) {
			result = true;
		} else if (list.isEmpty()) {
			result = true;
		}

		return result;
	}

	private boolean isEmptyStr(String str) {
		boolean result = false;

		if (str == null) {
			result = true;
		} else if (str.trim().isEmpty()) {
			result = true;
		}

		return result;
	}

	private boolean isLastUpdatedSameAsCreated() {
		Date lastUpdated = hazard.getLastUpdated();
		Date created = hazard.getCreated();

		return (lastUpdated != null) && (created != null)
				&& (lastUpdated.equals(created));
	}

	private int numberOfRowsInGeneralSection() {
		int numRows = 0;
		return numRows;
	}

	private int numberOfRowsInWhenSection() {
		int numRows = 0;

		return numRows;
	}

	private void removeFirstOfConsecutiveHeadings() {
		Set<CellRec> cellRecsToRemove = new HashSet<CellRec>();
		for (int i = 0; i < cellRecs.size(); i++) {
			CellRec thisrec = cellRecs.get(i);

			if (i < (cellRecs.size() - 2)) {
				CellRec nextrec = cellRecs.get(i + 1);

				if ((thisrec instanceof HeadingCellRec)
						&& (nextrec instanceof HeadingCellRec)) {
					cellRecsToRemove.add(thisrec);
				}
			}

			if (i == (cellRecs.size() - 1)) {
				if (thisrec instanceof HeadingCellRec) {
					cellRecsToRemove.add(thisrec);
				}
			}
		}

		cellRecs.removeAll(cellRecsToRemove);
	}

}

File: app/src/main/java/rod/bailey/trafficatnsw/hazard/details/HazardDetailsListAdapter.kt
package rod.bailey.trafficatnsw.hazard.details

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ListAdapter
import rod.bailey.trafficatnsw.R
import rod.bailey.trafficatnsw.app.ConfigSingleton
import rod.bailey.trafficatnsw.app.TrafficAtNSWApplication
import rod.bailey.trafficatnsw.common.ui.ListHeadingView_
import rod.bailey.trafficatnsw.hazard.data.XHazard
import rod.bailey.trafficatnsw.hazard.data.XLane
import rod.bailey.trafficatnsw.hazard.data.XProperties
import rod.bailey.trafficatnsw.hazard.details.ui.cellrec.*
import rod.bailey.trafficatnsw.hazard.details.ui.view.HtmlListItemView_
import rod.bailey.trafficatnsw.hazard.details.ui.view.LineListItemView_
import rod.bailey.trafficatnsw.hazard.details.ui.view.TextFieldListItemView_
import rod.bailey.trafficatnsw.hazard.ui.HazardListItemView_
import rod.bailey.trafficatnsw.util.DateUtils
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject

class HazardDetailsListAdapter(private val ctx: Context, private val hazard: XHazard) : BaseAdapter(),
	ListAdapter {
	private val cellRecs = LinkedList<CellRec>()

	@Inject
	lateinit var config: ConfigSingleton

	init {
		TrafficAtNSWApplication.graph.inject(this)

		// Title cell (always appears)
		addTitleCellRec()
		addHeadingCellRec(ctx.getString(R.string.hazard_details_heading_times))
		addCreateAndLastUpdatedCellRecs()
		addDurationCellRec()
		addHeadingCellRec(ctx.getString(R.string.hazard_details_heading_general))
		addAttendingCellRec()
		addRtaAdviceCellRec()
		addQueueLengthCellRec()
		addClosureCellRecs()
		addLaneCellRecs()
		addTrafficVolumeAndDelayCellRec()

		// Cells containing HTML
		addPublicTransportCellRec()
		addArrangementElementCellRecs()
		addOtherAdviceCellRec()

		// De-duplication of headings
		removeFirstOfConsecutiveHeadings()
	}

	override fun isEnabled(position: Int): Boolean = cellRecs[position] !is HeadingCellRec
	override fun areAllItemsEnabled(): Boolean = false

	private fun addArrangementElementCellRecs() {
		val props: XProperties? = hazard.properties
		if (props != null) {
			val elts = props.arrangementElements
			if ((elts != null) && (elts.size > 0)) {
				for (element in elts) {
					if (element.title != null) {
						addHeadingCellRec(element.title)
					}
					if (element.html != null) {
						addHtmlFieldCellRec(element.html)
					}
				}
			}
		}
	}

	private fun addAttendingCellRec() {
		val props: XProperties? = hazard.properties
		if (props != null) {
			val groups = props.attendingGroups
			if ((groups != null) && (groups.size > 0)) {
				val numAttendingGroups = groups.size
				val str = StringBuffer()

				for (i in 0..numAttendingGroups - 1) {
					val group = groups[i]

					if (!isEmptyStr(group)) {
						str.append(group)
					}

					if (i < numAttendingGroups - 1) {
						str.append(", ")
					}
				} // for i

				val strstr = str.toString()
				if (!isEmptyStr(strstr)) {
					val cellrec = TextFieldCellRec(ctx.getString(R.string.hazard_details_field_label_attending), strstr)
					cellRecs.add(cellrec)
				}
			}
		}
	}

	private fun addClosureCellRecs() {
		val props: XProperties? = hazard.properties
		if (props != null) {
			val periods = props.periods
			if (periods != null) {
				for (period in periods) {
					if ((XHazard.TOKEN_ROAD_CLOSURE == period.closureType) ||
						(XHazard.TOKEN_LANE_CLOSURE == period.closureType)) {
						val str = StringBuffer()

						if (!isEmptyStr(period.toDay)) {
							str.append(ctx.getString(R.string.hazard_details_closure_from_date_to_date),
									   period.fromDay, period.toDay)
						} else {
							str.append(period.fromDay)
						}

						if (!isEmptyStr(period.startTime)) {
							str.append(" ")
							if (!isEmptyStr(period.finishTime)) {
								str.append(ctx.getString(R.string.hazard_details_closure_from_time_to_time),
										   period.startTime, period.finishTime)
							}
						}

						if (!isEmptyStr(period.direction)) {
							str.append(ctx.getString(R.string.hazard_details_closure_direction_affected),
									   period.direction)
							str.append(" ")
						}

						val fieldValue = str.toString()
						val fieldName: String =
							if (XHazard.TOKEN_ROAD_CLOSURE == period.closureType) {
								ctx.getString(R.string.hazard_details_field_label_roads_closed)
							} else {
								ctx.getString(R.string.hazard_details_field_label_lanes_closed)
							}
						val cellRec = TextFieldCellRec(fieldName, fieldValue)
						cellRecs.add(cellRec)
					}
				} // for
			}
		}
	}

	private fun addCreateAndLastUpdatedCellRecs() {
		// First CellRec is for "Created" date/time
		val props: XProperties? = hazard.properties
		if ((props != null) && (props.created != null)) {
			val createdStr = DateUtils.relativeDateAndTime(
				Date(props.created), capitalize = true)
			val createdCellRec = TextFieldCellRec(
				fieldName = ctx.getString(R.string.hazard_details_field_label_when_started),
				fieldValue = createdStr)
			cellRecs.add(createdCellRec)
		}

		// Second CellRec is for "Last updated" date/time. But we only create a
		// 'last updated' cellrec if it will show a different date to the
		// 'created' date.
		if ((props != null) && (!isLastUpdatedSameAsCreated) && (props.lastUpdated != null)) {
			val lastUpdatedStr = DateUtils.relativeDateAndTime(
				Date(props.lastUpdated), capitalize = true)
			val lastUpdatedCellRec = TextFieldCellRec(
				fieldName = ctx.getString(R.string.hazard_details_field_label_when_last_checked),
				fieldValue = lastUpdatedStr)
			cellRecs.add(lastUpdatedCellRec)
		}
	}

	private fun addDurationCellRec() {
		val props: XProperties? = hazard.properties
		if (props != null) {
			if (props.start != null && props.end != null) {
				val formatString = StringBuffer(config.durationTimeFormat())
				val sdf = SimpleDateFormat(formatString.toString(), Locale.ENGLISH)
				val startStr = sdf.format(props.start)
				val endStr = sdf.format(props.end)
				val startToEnd = StringBuffer()
				startToEnd.append(startStr)
				startToEnd.append(" - ")
				startToEnd.append(endStr)
				val cellrec = TextFieldCellRec(ctx.getString(R.string.hazard_details_field_label_duration),
											   startToEnd.toString())
				cellRecs.add(cellrec)
			}
		}
	}

	private fun addHeadingCellRec(heading: String) = cellRecs.add(HeadingCellRec(heading))
	private fun addHtmlFieldCellRec(html: String) = cellRecs.add(HtmlFieldCellRec(html))

	private fun createLaneCellRecAffected(lane: XLane): CellRec {
		val line: String =
			if (XHazard.TOKEN_EXTENT_BOTH == lane.affectedDirection) {
				ctx.getString(R.string.hazard_details_field_value_extent_both)
			} else {
				ctx.getString(R.string.hazard_details_field_value_extent_one,
							  lane.affectedDirection)
			}
		return LineCellRec(line)
	}

	/**
	 * Form a string of form "X of Y <DIR>bound lanes closed"
	 * e.g. "2 of 3 eastbound lanes closed".
	 * NOTE: Special case where X and Y are blank implies that both
	 * X and Y are "1", but instead of saying "1 of 1 <DIR>bound lanes closed"
	 * we say "<DIR>bound lane closed".
	 */
	private fun createLaneCellRecLanesClosed(lane: XLane): CellRec {
		val str = StringBuffer()

		if (isEmptyStr(lane.closedLanes) && isEmptyStr(lane.numberOfLanes)) {
			str.append(ctx.getString(R.string.hazard_details_field_value_extent_one,
									 lane.affectedDirection))
		} else {
			str.append(ctx.getString(R.string.hazard_details_field_value_one_of_n_lanes_closed,
									 lane.closedLanes, lane.numberOfLanes,
									 lane.affectedDirection?.toLowerCase() ?: ""))
		}

		// Suffix "(<description>)" e.g. "(right turn lane)"
		if (!isEmptyStr(lane.description)) {
			str.append(" ")
			str.append(ctx.getString(R.string.hazard_details_parenthetical_lane_description,
									 lane.description?.toLowerCase() ?: ""))
		}

		str.append(".")
		return LineCellRec(str.toString())
	}

	private fun createLaneCellRecLaneClosed(lane: XLane): CellRec {
		val str = StringBuffer()

		if (XHazard.TOKEN_EXTENT_BOTH == lane.affectedDirection) {
			str.append(ctx.getString(R.string.hazard_details_road_closed_in_both_directions,
									 lane.roadType))
		} else {
			str.append(ctx.getString(R.string.hazard_details_road_closed_in_one_direction,
									 lane.roadType,
									 lane.affectedDirection?.toLowerCase() ?: ""))
		}
		return LineCellRec(str.toString())
	}

	private fun addLaneCellRecs() {
		val props: XProperties? = hazard.properties
		if ((props != null) && (props.roads != null) && (props.roads.size > 0)) {
			val firstRoad = props.roads[0]

			if (!isEmptyList(firstRoad.impactedLanes)) {
				if (firstRoad.impactedLanes != null) {
					for (lane in firstRoad.impactedLanes) {
						if (!isEmptyStr(lane.extent)) {
							when (lane.extent) {
								XHazard.TOKEN_EXTENT -> cellRecs.add(createLaneCellRecAffected(lane))
								XHazard.TOKEN_LANES_CLOSED -> cellRecs.add(createLaneCellRecLanesClosed(lane))
								XHazard.TOKEN_EXTENT_LANE_CLOSED -> cellRecs.add(createLaneCellRecLaneClosed(lane))
							}
						}
					}
				}
			}
		}
	}

	private fun addOtherAdviceCellRec() {
		val props: XProperties? = hazard.properties
		if ((props != null) && (props.otherAdvice != null) && (!isEmptyStr(props.otherAdvice))) {
			addHeadingCellRec(ctx.getString(R.string.hazard_details_heading_other_advice))
			addHtmlFieldCellRec(props.otherAdvice)
		}
	}

	private fun addPublicTransportCellRec() {
		val props: XProperties? = hazard.properties
		if ((props != null) && !isEmptyStr(props.publicTransport)) {
			addHeadingCellRec(ctx.getString(R.string.hazard_details_heading_public_transport))
			if (props.publicTransport != null) {
				addHtmlFieldCellRec(props.publicTransport)
			}
		}
	}

	private fun addQueueLengthCellRec() {
		val props: XProperties? = hazard.properties
		if ((props != null) && (props.roads != null) && !isEmptyList(props.roads)) {
			val firstRoad = props.roads[0]

			if ((firstRoad.queueLength != null) && (firstRoad.queueLength > 0)) {
				val fieldName = ctx.getString(R.string.hazard_details_field_label_queues)
				val fieldValue: String = ctx.getString(R.string.hazard_details_field_value_queues,
													   firstRoad.queueLength)
				val cellrec = TextFieldCellRec(fieldName, fieldValue)
				cellRecs.add(cellrec)
			}
		}
	}

	private fun addRtaAdviceCellRec() {
		val props: XProperties? = hazard.properties
		if (props != null) {
			if (!isEmptyStr(props.adviceA) || !isEmptyStr(props.adviceB)) {
				val value = StringBuffer()
				val fieldName = ctx.getString(R.string.hazard_details_rta_advice)

				if (!isEmptyStr(props.adviceA)) {
					value.append(props.adviceA)

					if (!isEmptyStr(props.adviceB)) {
						value.append(", ")
					}
				}

				if (!isEmptyStr(props.adviceB)) {
					value.append(props.adviceB)
				}
				val cellRec = TextFieldCellRec(fieldName, fieldValue = value.toString())
				cellRecs.add(cellRec)
			}
		}
	}

	private fun addTitleCellRec() = cellRecs.add(TitleCellRec(hazard))

	/**
	 * "Traffic Volume" and "Delay" - adds a single LineCellRec containing a
	 * summary of both like "Heavy traffic. Significan delays" where the stings
	 * "Heavy" and "Significant" are taken from the JSON file.
	 */
	private fun addTrafficVolumeAndDelayCellRec() {
		val props: XProperties? = hazard.properties
		if ((props != null) && (props.roads != null) && !isEmptyList(props.roads)) {
			val firstRoad = props.roads[0]
			val str = StringBuffer()

			if (!isEmptyStr(firstRoad.trafficVolume)) {
				str.append(ctx.getString(R.string.hazard_details_traffic_volume, firstRoad.trafficVolume))
			}

			if (!isEmptyStr(firstRoad.delay)) {
				if (str.length > 1) {
					str.append(" ")
				}

				str.append(ctx.getString(R.string.hazard_details_traffic_delays, firstRoad.delay))
			}

			if (str.length > 1) {
				val rec = LineCellRec(str.toString())
				cellRecs.add(rec)
			}
		}
	}

	override fun getCount(): Int = cellRecs.size
	override fun getItem(position: Int): Any = cellRecs[position]
	override fun getItemId(position: Int): Long = position.toLong()

	override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
		val result: View
		val ctx = parent.context
		val cellRec: CellRec = cellRecs[position]

		result = when (cellRec) {
			is TitleCellRec -> HazardListItemView_.build(ctx, cellRec.hazard, false, false)
			is HeadingCellRec -> ListHeadingView_.build(ctx, cellRec.heading)
			is LineCellRec -> LineListItemView_.build(ctx, cellRec.line)
			is TextFieldCellRec -> TextFieldListItemView_.build(ctx, cellRec.fieldName, cellRec.fieldValue)
			is HtmlFieldCellRec -> HtmlListItemView_.build(ctx, cellRec.fieldHtml)
			else -> View(ctx)
		}

		return result
	}

	private fun isEmptyList(list: List<*>?): Boolean {
		var result = false

		if (list == null) {
			result = true
		} else if (list.isEmpty()) {
			result = true
		}

		return result
	}

	private fun isEmptyStr(str: String?): Boolean {
		var result = false

		if (str == null) {
			result = true
		} else if (str.trim { it <= ' ' }.isEmpty()) {
			result = true
		}

		return result
	}

	private val isLastUpdatedSameAsCreated: Boolean
		get() {
			var result = false
			val props: XProperties? = hazard.properties

			if (props != null) {
				val lastUpdated: Long? = props.lastUpdated
				val created: Long? = props.created

				result = (lastUpdated != null) && (created != null) && (lastUpdated == created)
			}

			return result
		}

	private fun removeFirstOfConsecutiveHeadings() {
		val cellRecsToRemove = HashSet<CellRec>()
		for (i in cellRecs.indices) {
			val thisrec = cellRecs[i]

			if (i < cellRecs.size - 2) {
				val nextrec = cellRecs[i + 1]

				if (thisrec is HeadingCellRec && nextrec is HeadingCellRec) {
					cellRecsToRemove.add(thisrec)
				}
			}

			if (i == cellRecs.size - 1) {
				if (thisrec is HeadingCellRec) {
					cellRecsToRemove.add(thisrec)
				}
			}
		}

		cellRecs.removeAll(cellRecsToRemove)
	}
}
